pkg/osbuild/librepo_source.go

```diff
@@ -0,0 +1,120 @@
+}
+
+func NewLibrepoSource() *LibrepoSource {
+	return &LibrepoSource{
+		Items: make(map[string]*librepoSourceItem),
+		Options: &librepoSourceOptions{
+			Mirrors: make(map[string]*librepoSourceMirror),
+		},
+	}
+}
+
+// AddPackage adds the given *depsolved* pkg to the downloading. It
+// needs the *depsovled* repoConfig so that the repoID of the two can
+// be matched up
+func (source *LibrepoSource) AddPackage(pkg rpmmd.PackageSpec, repos []rpmmd.RepoConfig) error {
+	pkgRepo := findRepoById(repos, pkg.RepoID)
+	if pkgRepo == nil {
+		return fmt.Errorf("cannot find repo-id %v for %v in %+v", pkg.RepoID, pkg.Name, repos)
+	}
+	if _, ok := source.Options.Mirrors[pkgRepo.Id]; !ok {
+		mirror, err := mirrorFromRepo(pkgRepo)
+		if err != nil {
+			return err
+		}
+		source.Options.Mirrors[pkgRepo.Id] = mirror
+	}
+	mirror := source.Options.Mirrors[pkgRepo.Id]
+	// XXX: should we error here if one package requests IgnoreSSL
+	// and one does not for the same mirror?
+	if pkg.IgnoreSSL {
+		mirror.Insecure = true
+	}
+	if pkg.Secrets == "org.osbuild.rhsm" {
+		mirror.Secrets = &URLSecrets{
+			Name: "org.osbuild.rhsm",
+		}
+	} else if pkg.Secrets == "org.osbuild.mtls" {
+		mirror.Secrets = &URLSecrets{
+			Name: "org.osbuild.mtls",
+		}
+	}
+
+	item := &librepoSourceItem{
+		Path:     pkg.Path,
+		MirrorID: pkgRepo.Id,
+	}
+	source.Items[pkg.Checksum] = item
+	return nil
+}
+
+func (LibrepoSource) isSource() {}
+
+type librepoSourceItem struct {
+	Path     string `json:"path"`
+	MirrorID string `json:"mirror"`
+}
+
+func findRepoById(repos []rpmmd.RepoConfig, repoID string) *rpmmd.RepoConfig {
+	for _, repo := range repos {
+		if repo.Id == repoID {
+			return &repo
+		}
+	}
+
+	return nil
+}
+
+func mirrorFromRepo(repo *rpmmd.RepoConfig) (*librepoSourceMirror, error) {
+	switch {
+	case repo.Metalink != "":
+		return &librepoSourceMirror{
+			URL:  repo.Metalink,
+			Type: "metalink",
+		}, nil
+	case repo.MirrorList != "":
+		return &librepoSourceMirror{
+			URL:  repo.MirrorList,
+			Type: "mirrorlist",
+		}, nil
+	case len(repo.BaseURLs) > 0:
+		return &librepoSourceMirror{
+			// XXX: should we pick a random one instead?
+			URL:  repo.BaseURLs[0],
+			Type: "baseurl",
+		}, nil
+	}
+
+	return nil, fmt.Errorf("cannot find metalink, mirrorlist or baseurl for %+v", repo)
+}
+
+// librepoSourceOptions are the JSON options for source org.osbuild.librepo
+type librepoSourceOptions struct {
+	Mirrors map[string]*librepoSourceMirror `json:"mirrors"`
+}
+
+type librepoSourceMirror struct {
+	URL  string `json:"url"`
+	Type string `json:"type"`
+
+	Insecure bool        `json:"insecure,omitempty"`
+	Secrets  *URLSecrets `json:"secrets,omitempty"`
+
+	// XXX: should we expose those? if so we need a way to set them,
+	// current this is done in manifest.GenSources which cannot take
+	// options.
+	// MaxParallels  *int `json:"max-parallels,omitempty"`
+	// FastestMirror bool `json:"fastest-mirror,omitempty"`
```

Comment:
Expose them anyway.  I think it's better if the bindings/wrappers in this package reflect the osbuild stages even when we don't use specific options.  It's always possible to import the `osbuild` package and create stages with their options manually, without using a helper function or even the `NewXStage()` initialiser.

I'd also make `librepoSourceMirror` an exported type.